src/GraphQL/Execution/DocumentExecuter.cs

```diff
@@ -71,8 +71,14 @@ public async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options)
                 {
                     using (metrics.Subject("schema", "Initializing schema"))
                     {
-                        options.FieldMiddleware.ApplyTo(options.Schema);
-                        options.Schema.Initialize();
+                        lock (options.Schema)
```

Comment:
Have you considered using double-check locking here? I have no figures to hand but as I understand it, adding another `if (!options.Schema.Initialized)` outside of the lock would be more efficient. There is a small chance of race condition so the second check is performed inside the scope of the lock (as you current have it).

Reply:
Just look a few lines above.

Reply:
Yeah, we commented at the same time